[libc] added newhdrgen python script and class file

aaryanshukla opened the pull request:
python script uses yaml and classes to generate c headers
header.py is only the main class file, the rest will be in another pr
more files to be added in multiple prs

aaryanshukla committed:
[libc] added newhdrgen python script and class file

python script uses yaml and classes to generate c headers
header.py is only the main class file, the rest will be in another pr
more files to be added in multiple prs

lntue commented on libc/newhdrgen/header.py:
```diff
@@ -0,0 +1,59 @@
+# class_implementation/classes/header.py
```
Add license header and a bit more detailed description in the comments.

aaryanshukla replied:
just pushed comments and license header.

jhuber6 commented on libc/newhdrgen/yaml_to_classes.py:
```diff
@@ -0,0 +1,113 @@
+
+    main(args.yaml_file, args.h_def_file, args.output_dir)
+
+# Example Command Line Arg: python3 yaml_to_classes.py yaml/stdc_stdbit.yaml h_def/stdbit.h.def --output_dir output
```
This should probably just be the output when called with no inputs or `--help` or something.

aaryanshukla committed:
added comments and license header

aaryanshukla committed:
code formatted fixed

nickdesaulniers commented on libc/newhdrgen/yaml_to_classes.py:
```diff
@@ -0,0 +1,142 @@
+    header_str = str(header)
+    final_header_content = fill_public_api(header_str, h_def_content)
+
+    output_file_name = os.path.basename(h_def_file).replace(".def", "")
+    output_file_path = os.path.join(output_dir, output_file_name)
```
Python now has a `Path` class. Please use that instead of `os.path`.
https://docs.python.org/3/library/pathlib.html

RoseZhang03 commented on libc/newhdrgen/header.py:
```diff
@@ -0,0 +1,59 @@
+        self.macros = []
+        self.types = []
+        self.enumerations = []
+        self.objects = []
```
objects are referring to the Object specs found in stdc.td and posix.td

aaryanshukla committed:
addressed comments regarding header.py and os lib

added license header to header.py and removed comments
used pathlib instead of os in yaml_to_classes

lntue commented on libc/newhdrgen/yaml_to_classes.py:
```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python
+#
+# ====- Generate headers for libc functions  ------------*- python -*--==#
```
alignment at the end

lntue commented on libc/newhdrgen/header.py:
```diff
@@ -0,0 +1,65 @@
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# ==-------------------------------------------------------------------------==#
+class HeaderFile:
```
give them 1-2 empty lines

aaryanshukla committed:
fixed alignment and spacing issues

aaryanshukla committed:
changed one of the class names for imports

nickdesaulniers commented on libc/newhdrgen/yaml_to_classes.py:
```diff
@@ -0,0 +1,144 @@
+    Returns:
+        The final header content with the public API filled in.
+    """
+    return re.sub(r"%%public_api\(\)", header_str, h_def_content)
```
Do you even need regular expressions here? Will https://docs.python.org/3/library/stdtypes.html#str.replace not work here?

aaryanshukla replied:
`return h_def_content.replace("%%public_api()", header_str)` works, regular expressions is no longer neeeded.

nickdesaulniers replied:
Consider using the `count` parameter of the `replace` method. `%%public_api()` should only appear once per .h.def file. Once we've found it, we can stop.

aaryanshukla replied:
Will be adding `return h_def_content.replace("%%public_api()", header_str, 1)` to the next commit.

aaryanshukla committed:
replace re.sub with str.replace

michaelrj-google commented on libc/newhdrgen/yaml_to_classes.py:
```diff
@@ -0,0 +1,144 @@
+
+from pathlib import Path
+from header import HeaderFile
+from class_implementation.classes.macro import Macro
```
you should make sure to land a patch adding these classes before landing this patch. Ideally the code in the repository should always work, even if it's not complete yet.

RoseZhang03 replied:
That patch should be ready to land today.

michaelrj-google reviewed (approved):
LGTM now that the other patch has landed
